File: tests/python/test_audio_volume.py
"""
Audio System Volume Control Tests
Tests volume control, clamping, and pitch validation.
"""

from test_common import *

def test_master_volume():
    """Test: Master volume control"""
    print("Test: Master volume control... ", end="", flush=True)
    audio = audio_py.AudioManager.get_instance()
    audio.initialize()
    
    audio.set_master_volume(0.5)
    volume = audio.get_master_volume()
    assert abs(volume - 0.5) < 0.01, f"Expected 0.5, got {volume}"
    
    audio.set_master_volume(1.0)
    assert abs(audio.get_master_volume() - 1.0) < 0.01, "Master volume should be 1.0"
    
    audio.set_master_volume(0.0)
    assert abs(audio.get_master_volume() - 0.0) < 0.01, "Master volume should be 0.0"
    
    # Test volume clamping
    audio.set_master_volume(-0.5)
    assert audio.get_master_volume() >= 0.0, "Negative master volume should be clamped"
    
    audio.set_master_volume(2.0)
    assert abs(audio.get_master_volume() - 1.0) < 0.01, "Master volume > 1.0 should be clamped"
    
    audio.set_master_volume(1.0)  # Reset
    audio.shutdown()
    print("PASS")

def test_volume_clamping():
    """Test: Volume clamping for all methods"""
    print("Test: Volume clamping... ", end="", flush=True)
    audio = audio_py.AudioManager.get_instance()
    audio.initialize()
    
    # Test group volume clamping
    group = audio.create_group("test")
    audio.set_group_volume(group, -1.0)
    assert audio.get_group_volume(group) >= 0.0, "Negative group volume should be clamped"
    
    audio.set_group_volume(group, 5.0)
    assert abs(audio.get_group_volume(group) - 1.0) < 0.01, "Group volume > 1.0 should be clamped"
    
    # Test sound volume clamping
    if sound_exists("digital_base.wav"):
        sound = audio.load_sound(get_sound_path("digital_base.wav"))
        audio.set_sound_volume(sound, -0.3)
        assert True, "Negative sound volume should be clamped (no crash)"
        
        audio.set_sound_volume(sound, 10.0)
        assert True, "Sound volume > 1.0 should be clamped (no crash)"
        
        # Test layer volume clamping
        track = audio.create_track()
        audio.add_layer(track, "layer1", get_sound_path("digital_base.wav"))
        audio.set_layer_volume(track, "layer1", -0.5)
        assert True, "Negative layer volume should be clamped (no crash)"
        
        audio.set_layer_volume(track, "layer1", 2.5)
        assert True, "Layer volume > 1.0 should be clamped (no crash)"
        
        audio.destroy_track(track)
        audio.destroy_sound(sound)
    
    audio.destroy_group(group)
    audio.shutdown()
    print("PASS")

def test_pitch_validation():
    """Test: Pitch validation and clamping"""
    print("Test: Pitch validation... ", end="", flush=True)
    audio = audio_py.AudioManager.get_instance()
    audio.initialize()
    
    if sound_exists("digital_base.wav"):
        sound = audio.load_sound(get_sound_path("digital_base.wav"))
        
        # Test negative pitch (should be clamped)
        audio.set_sound_pitch(sound, -1.0)
        assert True, "Negative pitch should be clamped (no crash)"
        
        # Test zero pitch (should be clamped)
        audio.set_sound_pitch(sound, 0.0)
        assert True, "Zero pitch should be clamped (no crash)"
        
        # Test very high pitch (should be clamped)
        audio.set_sound_pitch(sound, 100.0)
        assert True, "Very high pitch should be clamped (no crash)"
        
        # Test valid pitch values
        audio.set_sound_pitch(sound, 0.5)
        assert True, "Valid pitch 0.5 should work"
        
        audio.set_sound_pitch(sound, 1.0)
        assert True, "Valid pitch 1.0 should work"
        
        audio.set_sound_pitch(sound, 2.0)
        assert True, "Valid pitch 2.0 should work"
        
        audio.destroy_sound(sound)
    else:
        print("SKIP (no sound file) ", end="")
    
    audio.shutdown()
    print("PASS")

def run_all_tests():
    """Run all volume tests"""
    print("=== Python Audio Volume Control Tests ===\n")
    
    tests = [
        test_master_volume,
        test_volume_clamping,
        test_pitch_validation,
    ]
    
    passed = 0
    failed = 0
    
    for test in tests:
        try:
            test()
            passed += 1
        except AssertionError as e:
            print(f"FAIL - {e}")
            failed += 1
        except Exception as e:
            print(f"ERROR - {e}")
            failed += 1
    
    print(f"\n=== Results: {passed} passed, {failed} failed ===")
    return failed == 0

if __name__ == "__main__":
    success = run_all_tests()
    sys.exit(0 if success else 1)

File: tests/cpp/test_audio_volume.cpp
/**
 * Audio System Volume Control Tests
 * 
 * Tests volume control, clamping, and pitch validation.
 */

#include "test_common.h"
#include "audio_manager.h"
#include <cmath>
#include <algorithm>

using namespace audio;

void test_master_volume() {
    TEST("Master Volume Control")
    
    auto& audio = AudioManager::GetInstance();
    
    audio.SetMasterVolume(0.5f);
    ASSERT(std::abs(audio.GetMasterVolume() - 0.5f) < 0.01f, "Master volume should be 0.5")
    
    audio.SetMasterVolume(1.0f);
    ASSERT(std::abs(audio.GetMasterVolume() - 1.0f) < 0.01f, "Master volume should be 1.0")
    
    audio.SetMasterVolume(0.0f);
    ASSERT(std::abs(audio.GetMasterVolume() - 0.0f) < 0.01f, "Master volume should be 0.0")
    
    // Test volume clamping
    audio.SetMasterVolume(-0.5f);
    ASSERT(audio.GetMasterVolume() >= 0.0f, "Negative master volume should be clamped to >= 0.0")
    
    audio.SetMasterVolume(2.0f);
    ASSERT(std::abs(audio.GetMasterVolume() - 1.0f) < 0.01f, "Master volume > 1.0 should be clamped to 1.0")
    
    audio.SetMasterVolume(1.0f); // Reset to normal
    
    END_TEST
}

void test_volume_clamping() {
    TEST("Volume Clamping for All Methods")
    
    auto& audio = AudioManager::GetInstance();
    
    // Test group volume clamping
    GroupHandle group = audio.CreateGroup();
    audio.SetGroupVolume(group, -1.0f);
    ASSERT(audio.GetGroupVolume(group) >= 0.0f, "Negative group volume should be clamped")
    
    audio.SetGroupVolume(group, 5.0f);
    ASSERT(std::abs(audio.GetGroupVolume(group) - 1.0f) < 0.01f, "Group volume > 1.0 should be clamped")
    
    // Test sound volume clamping
    SoundHandle sound = audio.LoadSound(sound_dir + "/digital_base.wav");
    audio.SetSoundVolume(sound, -0.3f);
    ASSERT(true, "Negative sound volume should be clamped (no crash)")
    
    audio.SetSoundVolume(sound, 10.0f);
    ASSERT(true, "Sound volume > 1.0 should be clamped (no crash)")
    
    // Test layer volume clamping
    TrackHandle track = audio.CreateTrack();
    audio.AddLayer(track, "layer1", sound_dir + "/digital_base.wav");
    audio.SetLayerVolume(track, "layer1", -0.5f);
    ASSERT(true, "Negative layer volume should be clamped (no crash)")
    
    audio.SetLayerVolume(track, "layer1", 2.5f);
    ASSERT(true, "Layer volume > 1.0 should be clamped (no crash)")
    
    // Cleanup
    audio.DestroySound(sound);
    audio.DestroyTrack(track);
    audio.DestroyGroup(group);
    
    END_TEST
}

void test_pitch_validation() {
    TEST("Pitch Validation and Clamping")
    
    auto& audio = AudioManager::GetInstance();
    
    SoundHandle sound = audio.LoadSound(sound_dir + "/digital_base.wav");
    
    // Test negative pitch (should be clamped to minimum)
    audio.SetSoundPitch(sound, -1.0f);
    ASSERT(true, "Negative pitch should be clamped (no crash)")
    
    // Test zero pitch (should be clamped to minimum)
    audio.SetSoundPitch(sound, 0.0f);
    ASSERT(true, "Zero pitch should be clamped (no crash)")
    
    // Test very high pitch (should be clamped to maximum)
    audio.SetSoundPitch(sound, 100.0f);
    ASSERT(true, "Very high pitch should be clamped (no crash)")
    
    // Test valid pitch values
    audio.SetSoundPitch(sound, 0.5f);
    ASSERT(true, "Valid pitch 0.5 should work")
    
    audio.SetSoundPitch(sound, 1.0f);
    ASSERT(true, "Valid pitch 1.0 should work")
    
    audio.SetSoundPitch(sound, 2.0f);
    ASSERT(true, "Valid pitch 2.0 should work")
    
    audio.DestroySound(sound);
    
    END_TEST
}

int main(int argc, char* argv[]) {
    std::cout << "========================================" << std::endl;
    std::cout << "Audio System Volume Control Tests" << std::endl;
    std::cout << "========================================" << std::endl;
    
    sound_dir = get_sound_dir(argc, argv);
    std::cout << "Sound directory: " << sound_dir << std::endl;
    
    // Initialize audio system
    auto& audio = AudioManager::GetInstance();
    if (!audio.Initialize()) {
        std::cerr << "FATAL: Failed to initialize audio system" << std::endl;
        return 1;
    }
    
    // Run all tests
    test_master_volume();
    test_volume_clamping();
    test_pitch_validation();
    
    // Final shutdown
    audio.Shutdown();
    
    // Print results
    std::cout << "\n========================================" << std::endl;
    std::cout << "Test Results" << std::endl;
    std::cout << "========================================" << std::endl;
    std::cout << "Passed: " << tests_passed << std::endl;
    std::cout << "Failed: " << tests_failed << std::endl;
    std::cout << "Total:  " << (tests_passed + tests_failed) << std::endl;
    
    if (tests_failed == 0) {
        std::cout << "\nALL TESTS PASSED ✓" << std::endl;
        return 0;
    } else {
        std::cout << "\nSOME TESTS FAILED ✗" << std::endl;
        return 1;
    }
}
